File: backend/parsers.py
"""Python port of src/parsers.js.

Normalizes raw upstream payloads (Foursquare, Overpass, Reddit, GNews,
Guardian) into the consistent shapes the SQL ingestion layer and the
frontend components expect:

  Venue    { id, kind: 'restaurant' | 'place', source, name, categories,
             address, city, lat, lon, distance_m,
             amenity?, cuisine?, outdoor_seating?, takeaway? }
  FeedItem { id, kind: 'reddit' | 'gnews' | 'guardian', source,
             title, body, url, publishedAt (ISO 8601 str),
             ...type-specific fields preserved for the React components }

Every parser is defensive — a missing/malformed field never raises, it
just falls through to a safe default. Mirrors the JS implementation
(intentionally) so backend and frontend agree on the schema.
"""
from __future__ import annotations
from datetime import datetime, timezone
from typing import Any


# ── Coercion helpers ─────────────────────────────────────────

def _num(v: Any):
    if v is None:
        return None
    try:
        n = float(v)
        return n if n == n else None  # NaN guard
    except (TypeError, ValueError):
        return None


def _str(v: Any, fallback: str = "") -> str:
    if v is None:
        return fallback
    return v if isinstance(v, str) else str(v)


def _arr(v: Any) -> list:
    return v if isinstance(v, list) else []


def _osm_yes_no(v: Any):
    if v is True or v == "yes":
        return True
    if v is False or v == "no":
        return False
    return None


# ── Venue parsers ────────────────────────────────────────────

def parse_foursquare_restaurant(raw: dict) -> dict | None:
    """Foursquare Places v3 → normalized venue."""
    if not isinstance(raw, dict):
        return None
    loc = raw.get("location") or {}
    geo = (raw.get("geocodes") or {}).get("main") or {}
    cats = []
    for c in _arr(raw.get("categories")):
        if isinstance(c, str):
            cats.append(c)
        elif isinstance(c, dict) and c.get("name"):
            cats.append(_str(c["name"]))
    return {
        "id": _str(raw.get("fsq_id") or raw.get("id")),
        "kind": "restaurant",
        "source": "foursquare",
        "name": _str(raw.get("name")),
        "categories": cats,
        "address": _str(raw.get("address") or loc.get("address") or loc.get("formatted_address")),
        "city": raw.get("city") or loc.get("locality"),
        "lat": _num(raw.get("lat") if raw.get("lat") is not None else geo.get("latitude")),
        "lon": _num(raw.get("lon") if raw.get("lon") is not None else geo.get("longitude")),
        "distance_m": _num(raw.get("distance_m") if raw.get("distance_m") is not None else raw.get("distance")),
    }


def parse_overpass_place(raw: dict) -> dict | None:
    """Overpass element (node or way with `center` lifted) → normalized place."""
    if not isinstance(raw, dict):
        return None
    tags = raw.get("tags") or {}

    address = _str(raw.get("address"))
    if not address and (tags.get("addr:housenumber") or tags.get("addr:street")):
        address = f"{tags.get('addr:housenumber') or ''} {tags.get('addr:street') or ''}".strip()

    cuisine = raw.get("cuisine") if raw.get("cuisine") is not None else tags.get("cuisine")
    if isinstance(cuisine, str):
        cuisine = [s.strip() for s in cuisine.split(";") if s.strip()]
    else:
        cuisine = _arr(cuisine)

    raw_id = raw.get("id")
    if isinstance(raw_id, str) and raw_id.startswith("osm-"):
        node_id = raw_id
    elif raw_id is not None:
        node_id = f"osm-{raw.get('type') or 'node'}-{raw_id}"
    else:
        node_id = ""

    outdoor = bool(raw["outdoor_seating"]) if raw.get("outdoor_seating") is not None else _osm_yes_no(tags.get("outdoor_seating"))
    takeaway = bool(raw["takeaway"]) if raw.get("takeaway") is not None else _osm_yes_no(tags.get("takeaway"))

    return {
        "id": node_id,
        "kind": "place",
        "source": "overpass",
        "name": _str(raw.get("name") or tags.get("name")),
        "amenity": raw.get("amenity") or tags.get("amenity"),
        "categories": cuisine,
        "cuisine": cuisine,
        "address": address,
        "city": raw.get("city"),
        "lat": _num(raw.get("lat")),
        "lon": _num(raw.get("lon")),
        "distance_m": _num(raw.get("distance_m")),
        "outdoor_seating": outdoor,
        "takeaway": takeaway,
    }


# ── Feed-item parsers ────────────────────────────────────────

def parse_reddit_post(raw: dict) -> dict | None:
    """Reddit listing child {kind, data} OR an already-flat post → normalized feed item."""
    if not isinstance(raw, dict):
        return None
    inner = raw.get("data")
    r = inner if isinstance(inner, dict) and (inner.get("title") or inner.get("id")) else raw

    permalink_raw = _str(r.get("permalink"))
    if permalink_raw.startswith("http"):
        url = permalink_raw
    elif permalink_raw:
        url = f"https://reddit.com{permalink_raw}"
    else:
        url = ""

    author_raw = _str(r.get("author"))
    author = f"u/{author_raw}" if author_raw and not author_raw.startswith("u/") else author_raw

    subreddit = _str(r.get("subreddit_name_prefixed") or r.get("subreddit"))
    created_utc = _num(r.get("created_utc"))

    published_at = None
    if created_utc is not None:
        published_at = datetime.fromtimestamp(created_utc, tz=timezone.utc).isoformat().replace("+00:00", "Z")

    return {
        "id": _str(r.get("id")),
        "kind": "reddit",
        "source": "reddit",
        "title": _str(r.get("title")),
        "selftext": _str(r.get("selftext")),
        "body": _str(r.get("selftext")),
        "raw_text": _str(r.get("selftext")),
        "url": url,
        "permalink": url,
        "author": author,
        "subreddit": subreddit,
        "score": int(_num(r.get("score")) or 0),
        "num_comments": int(_num(r.get("num_comments")) or 0),
        "created_utc": created_utc,
        "publishedAt": published_at,
    }


def parse_gnews_article(raw: dict, idx: int = 0) -> dict | None:
    """GNews article → normalized feed item."""
    if not isinstance(raw, dict):
        return None
    source_obj = raw.get("source")
    source_name = _str(source_obj.get("name")) if isinstance(source_obj, dict) else ""
    return {
        "id": _str(raw.get("id") or f"gnews-{idx}-{raw.get('url') or raw.get('title') or ''}"),
        "kind": "gnews",
        "source": "gnews",
        "title": _str(raw.get("title")),
        "description": _str(raw.get("description")),
        "content": _str(raw.get("content")),
        "body": _str(raw.get("description") or raw.get("content")),
        "raw_text": _str(raw.get("description") or raw.get("content")),
        "url": _str(raw.get("url")),
        "publishedAt": _str(raw.get("publishedAt")) or None,
        # preserved nested shape for the NewsArticle component
        "source_obj": {"name": source_name},
    }


def parse_guardian_article(raw: dict, idx: int = 0) -> dict | None:
    """Guardian Content API result item → normalized feed item."""
    if not isinstance(raw, dict):
        return None
    fields = raw.get("fields") or {}
    trail = _str(raw.get("trail_text") or fields.get("trailText"))
    return {
        "id": _str(raw.get("id") or f"guardian-{idx}"),
        "kind": "guardian",
        "source": "guardian",
        "title": _str(raw.get("webTitle") or raw.get("title")),
        "section": _str(raw.get("sectionName") or raw.get("section")),
        "url": _str(raw.get("webUrl") or raw.get("url")),
        "publishedAt": _str(raw.get("webPublicationDate") or raw.get("publishedAt")) or None,
        "trail_text": trail,
        "body": _str(raw.get("trail_text") or fields.get("trailText") or fields.get("bodyText")),
        "raw_text": trail,
        "byline": _str(raw.get("byline") or fields.get("byline")),
    }


# ── Envelope parsers ─────────────────────────────────────────

SOURCE_TO_PARSER = {
    "foursquare": lambda raw, i=0: parse_foursquare_restaurant(raw),
    "overpass":   lambda raw, i=0: parse_overpass_place(raw),
    "reddit":     lambda raw, i=0: parse_reddit_post(raw),
    "gnews":      lambda raw, i: parse_gnews_article(raw, i),
    "guardian":   lambda raw, i: parse_guardian_article(raw, i),
}


def normalize_envelope(envelope: dict) -> dict:
    """Map every item in `envelope.data` through the matching parser.

    Returns the same envelope shape with normalized data and an accurate count.
    Items that can't be parsed (None) are dropped.
    """
    source = (envelope or {}).get("source")
    parser = SOURCE_TO_PARSER.get(source)
    if not envelope or parser is None:
        return {
            "source": source or "unknown",
            "ok": False,
            "count": 0,
            "data": [],
            "error": (envelope or {}).get("error") or "unknown source",
        }

    raw_list = envelope.get("data") if isinstance(envelope.get("data"), list) else []
    items = []
    for i, raw in enumerate(raw_list):
        parsed = parser(raw, i)
        if parsed is not None:
            items.append(parsed)

    return {
        "source": source,
        "ok": envelope.get("ok") is not False and not envelope.get("error"),
        "count": len(items),
        "data": items,
        "error": envelope.get("error"),
    }

File: src/parsers.js
// ─────────────────────────────────────────────────────────────
// src/parsers.js
// Normalizers that convert raw upstream API responses into the
// consistent structured format used throughout GoodEats.
//
// Consistent item shapes:
//   Venue    { id, kind: 'restaurant' | 'place', name, categories[],
//              address, city, lat, lon, distance_m,
//              amenity?, cuisine?, outdoor_seating?, takeaway? }
//   FeedItem { id, kind: 'reddit' | 'gnews' | 'guardian',
//              title, body, url, publishedAt (ISO 8601 string),
//              ...type-specific fields preserved for components }
//
// Consistent envelope (one per upstream source):
//   { source, ok, count, data, error }
//
// Every parser is defensive: a missing or malformed field never
// throws — it falls back to a safe default ("", null, []) so a
// broken upstream payload can't crash the UI.
// ─────────────────────────────────────────────────────────────

function num(v) {
  if (v == null) return null;
  const n = typeof v === 'number' ? v : parseFloat(v);
  return Number.isFinite(n) ? n : null;
}

function str(v, fallback = '') {
  if (v == null) return fallback;
  return typeof v === 'string' ? v : String(v);
}

function arr(v) {
  return Array.isArray(v) ? v : [];
}

function osmYesNo(v) {
  if (v === true || v === 'yes') return true;
  if (v === false || v === 'no') return false;
  return undefined;
}

// ─── Venue parsers ───────────────────────────────────────────

// Raw Foursquare Places v3 shape:
//   { fsq_id, name, distance,
//     location: { address, locality, formatted_address },
//     geocodes: { main: { latitude, longitude } },
//     categories: [{ id, name, icon }] }
// Also accepts the already-cleaned shape used in MOCK_FALLBACK.
export function parseFoursquareRestaurant(raw) {
  if (!raw || typeof raw !== 'object') return null;
  const loc = raw.location || {};
  const geo = (raw.geocodes && raw.geocodes.main) || {};
  const cats = arr(raw.categories)
    .map((c) => (typeof c === 'string' ? c : str(c && c.name)))
    .filter(Boolean);
  return {
    id: str(raw.fsq_id ?? raw.id),
    kind: 'restaurant',
    name: str(raw.name),
    categories: cats,
    address: str(raw.address ?? loc.address ?? loc.formatted_address),
    city: raw.city ?? loc.locality ?? null,
    lat: num(raw.lat ?? geo.latitude),
    lon: num(raw.lon ?? geo.longitude),
    distance_m: num(raw.distance_m ?? raw.distance),
  };
}

// Raw Overpass/OSM element shape:
//   { id, type: 'node'|'way', lat, lon,
//     tags: { name, amenity, cuisine: 'a;b;c',
//             'addr:housenumber', 'addr:street',
//             outdoor_seating: 'yes'|'no', takeaway: 'yes'|'no' } }
// Also accepts the already-cleaned shape from MOCK_FALLBACK.
export function parseOverpassPlace(raw) {
  if (!raw || typeof raw !== 'object') return null;
  const tags = raw.tags || {};

  let address = str(raw.address);
  if (!address && (tags['addr:housenumber'] || tags['addr:street'])) {
    address = `${tags['addr:housenumber'] || ''} ${tags['addr:street'] || ''}`.trim();
  }

  let cuisine = raw.cuisine != null ? raw.cuisine : tags.cuisine;
  if (typeof cuisine === 'string') {
    cuisine = cuisine.split(';').map((s) => s.trim()).filter(Boolean);
  } else {
    cuisine = arr(cuisine);
  }

  let id;
  if (raw.id != null && typeof raw.id === 'string' && raw.id.startsWith('osm-')) {
    id = raw.id;
  } else if (raw.id != null) {
    id = `osm-${raw.type || 'node'}-${raw.id}`;
  } else {
    id = '';
  }

  const outdoor = raw.outdoor_seating !== undefined
    ? !!raw.outdoor_seating
    : osmYesNo(tags.outdoor_seating);
  const takeaway = raw.takeaway !== undefined
    ? !!raw.takeaway
    : osmYesNo(tags.takeaway);

  return {
    id,
    kind: 'place',
    name: str(raw.name ?? tags.name),
    amenity: raw.amenity ?? tags.amenity ?? null,
    categories: cuisine,
    cuisine,
    address,
    city: raw.city ?? null,
    lat: num(raw.lat),
    lon: num(raw.lon),
    distance_m: num(raw.distance_m),
    outdoor_seating: outdoor,
    takeaway,
  };
}

// ─── Feed-item parsers ───────────────────────────────────────

// Raw Reddit listing children: { kind: 't3', data: { id, title, selftext,
//   author, subreddit_name_prefixed, score, num_comments, created_utc, permalink } }
// `permalink` is path-only — prefix with reddit.com when normalizing.
export function parseRedditPost(raw) {
  if (!raw || typeof raw !== 'object') return null;
  // Accept either a Reddit listing child {kind,data} or an already-flat post.
  const r = raw.data && (raw.data.title || raw.data.id) ? raw.data : raw;

  const permalinkRaw = str(r.permalink);
  const url = permalinkRaw.startsWith('http')
    ? permalinkRaw
    : (permalinkRaw ? `https://reddit.com${permalinkRaw}` : '');

  const authorRaw = str(r.author);
  const author = authorRaw && !authorRaw.startsWith('u/') ? `u/${authorRaw}` : authorRaw;

  const subreddit = str(r.subreddit_name_prefixed ?? r.subreddit);
  const createdUtc = num(r.created_utc);

  return {
    id: str(r.id),
    kind: 'reddit',
    title: str(r.title),
    selftext: str(r.selftext),
    body: str(r.selftext),
    url,
    permalink: url,
    author,
    subreddit,
    score: num(r.score) ?? 0,
    num_comments: num(r.num_comments) ?? 0,
    created_utc: createdUtc,
    publishedAt: createdUtc != null ? new Date(createdUtc * 1000).toISOString() : null,
  };
}

// Raw GNews article shape:
//   { title, description, content, url, image, publishedAt,
//     source: { name, url } }
// The `source: { name }` sub-object is preserved verbatim because
// the NewsArticle component reads `article.source?.name` directly.
export function parseGNewsArticle(raw, idx = 0) {
  if (!raw || typeof raw !== 'object') return null;
  const sourceName = raw.source && typeof raw.source === 'object' ? str(raw.source.name) : '';
  return {
    id: str(raw.id ?? `gnews-${idx}-${raw.url || raw.title || ''}`),
    kind: 'gnews',
    title: str(raw.title),
    description: str(raw.description),
    content: str(raw.content),
    body: str(raw.description || raw.content),
    url: str(raw.url),
    publishedAt: str(raw.publishedAt) || null,
    source: { name: sourceName }, // preserved for component compatibility
  };
}

// Raw Guardian Content API shape:
//   { response: { results: [{ id, webTitle, sectionId, sectionName,
//       webUrl, webPublicationDate, fields: { trailText, byline, bodyText } }] } }
// `parseGuardianArticle` operates on a single result item.
export function parseGuardianArticle(raw, idx = 0) {
  if (!raw || typeof raw !== 'object') return null;
  const fields = raw.fields || {};
  return {
    id: str(raw.id ?? `guardian-${idx}`),
    kind: 'guardian',
    title: str(raw.webTitle ?? raw.title),
    section: str(raw.sectionName ?? raw.section),
    url: str(raw.webUrl ?? raw.url),
    publishedAt: str(raw.webPublicationDate ?? raw.publishedAt) || null,
    trail_text: str(raw.trail_text ?? fields.trailText),
    body: str(raw.trail_text ?? fields.trailText ?? fields.bodyText),
    byline: str(raw.byline ?? fields.byline),
  };
}

// ─── Envelope parsers ────────────────────────────────────────

const SOURCE_TO_PARSER = {
  foursquare: parseFoursquareRestaurant,
  overpass: parseOverpassPlace,
  reddit: parseRedditPost,
  gnews: parseGNewsArticle,
  guardian: parseGuardianArticle,
};

// Convert one per-source envelope `{ source, ok, count, data, error }`
// into the normalized form. `data` is mapped through the appropriate
// item parser. If `ok` is false or `data` is missing, returns an empty
// envelope with the error preserved.
export function parseSourceEnvelope(envelope) {
  const source = envelope && envelope.source;
  const parser = SOURCE_TO_PARSER[source];
  if (!envelope || !parser) {
    return {
      source: str(source) || 'unknown',
      ok: false,
      count: 0,
      data: [],
      error: envelope?.error ?? 'unknown source',
    };
  }
  const rawList = arr(envelope.data);
  const items = rawList.map((item, i) => parser(item, i)).filter(Boolean);
  return {
    source,
    ok: envelope.ok !== false && !envelope.error,
    count: items.length,
    data: items,
    error: envelope.error ?? null,
  };
}

// Top-level normalizer for the `/api/all` response. Walks every
// known source key on the response and produces a same-keyed
// object of normalized envelopes. Unknown keys are dropped.
export function parseApiResponse(raw) {
  if (!raw || typeof raw !== 'object') return null;
  const out = {};
  const sourceKeys = {
    restaurants: 'foursquare',
    places: 'overpass',
    reddit: 'reddit',
    news: 'gnews',
    articles: 'guardian',
  };
  for (const [key, expectedSource] of Object.entries(sourceKeys)) {
    const env = raw[key];
    if (!env) continue;
    // If the envelope is missing `source`, infer it from the key
    // so a backend that only sends `data: [...]` per slot still works.
    const withSource = env.source ? env : { ...env, source: expectedSource };
    out[key] = parseSourceEnvelope(withSource);
  }
  return out;
}
